ultravox/data/datasets.py

```diff
@@ -1153,17 +1207,34 @@ def __iter__(self):
         return (self._process(sample) for sample in self._dataset)
 
 
-class Range(data.IterableDataset):
+class Range(SizedIterableDataset):
     """Limits the number of samples from another dataset."""
 
     def __init__(
-        self, dataset: data.IterableDataset, num_samples: Optional[int] = None
+        self, dataset: SizedIterableDataset, num_samples: Optional[int] = None
     ) -> None:
+        super().__init__(dataset, len(dataset))
         self._dataset = dataset
         self._num_samples = num_samples
 
+        if self._num_samples is not None and self._num_samples > len(self._dataset):
+            # Issuing a warning here instead of raising an error to accomodate for specific classes of VoiceDataset
+            # Once we migrate entirely to GenericVoiceDataset, we can raise an error here.
+            warnings.warn("num_samples is greater than total_samples.")
+
     def __iter__(self):
-        for i, sample in enumerate(self._dataset):
-            if self._num_samples is not None and i >= self._num_samples:
+        count = 0
+        limit = (
+            self._num_samples if self._num_samples is not None else len(self._dataset)
+        )
+
+        for item in self._dataset:
+            if count >= limit:
                 break
-            yield sample
+            yield item
+            count += 1
```

Comment:
I see two issues here:
1. this is breaking the original interface of Range when num_samples is not specified and I'm not sure why. And in this case any dataset whose length is not specified correctly is gonna be incorrectly subsampled (in many cases to 0 samples, i.e. deleted)

2. I don't see why the main code was changed here. You could simply do `num_samples = ...` and keep the original code.

Reply:
Yes, good catch, I assumed the len() invariant incorrectly here.